students/prgrover/lesson09/mailroom.py

```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/env python3
+
+import textwrap
+import sys
+
+
+class Donor:
+    '''
+    Base class that encapsulates all donor information per donor.
+
+    For example, donor name, donation amount, total donation amount, numeber of donations, etc.
+    '''
+
+    def __init__(self, donor_name):
+        self.donor_name = donor_name
+        self.donations = []
+
+    @property
+    def donation_total(self):
+        return sum(self.donations)
+
+    @property
+    def donation_number(self):
+        return len(self.donations)
+
+    @property
+    def donation_average(self):
+        return self.donation_total / self.donation_number
+
+    def add_donation(self, donation_amount):
+        self.donations.append(donation_amount)
+
+
+class DonorDirectory:
+    '''
+    Encapsulates "database type" functions (e.g. insert, query, list, etc.)
+    '''
+
+    def __init__(self):
+        self.donor_list = []
+
+    def add_donor(self, donor_name, donation_amount):
+        donor = self.find_donor(donor_name)
+        if donor:
+            donor.add_donation(donation_amount)
+        else:
+            donor = Donor(donor_name)
+            donor.add_donation(donation_amount)
+            self.donor_list.append(donor)
+
+    def find_donor(self, donor_name):
+        for donor in self.donor_list:
+            if donor.donor_name == donor_name:
+                return donor
+
+    def list_donors(self):
+        output = '\n'.join([donor.donor_name for donor in self.donor_list])
+        if len(output) == 0:
+            output = "\nCurrently there are no donor records."
+        return output
+
+    def create_report(self):
+        donor_name = [donor.donor_name for donor in self.donor_list]
+        total_given = [donor.donation_total for donor in self.donor_list]
+        num_donations = [donor.donation_number for donor in self.donor_list]
+        average_donation = [donor.donation_average for donor in self.donor_list]
+        sorted_donor_list = sorted(zip(donor_name, total_given, num_donations, average_donation), key=lambda x: int(x[1]), reverse=True)
+
+        print("{:30s} | {:11s} | {:9s} | {:12s}".format("Donor Name", "Total Given", "Num Donations", "Average Donation"))
+        print("-" * 79)
+
+        for index in range(len(sorted_donor_list)):
+            row = sorted_donor_list[index]
+            print("{:30s}  ${:11.2f}   {:12d}    ${:12.2f}".format(*row))
+
+
+    def send_mail(self, donor_name, donation_amount):
```

Comment:
suggested name -> `get_letter` and this would be better to live in `Donor` class since it is dealing with single donor only.